Sources/SwiftGRPCNIO/LengthPrefixedMessageReader.swift

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+
+/// This class reads and decodes length-prefixed gRPC messages.
+///
+/// Messages are expected to be in the following format:
+/// - compression flag: 0/1 as a 1-byte unsigned integer,
+/// - message length: length of the message as a 4-byte unsigned integer,
+/// - message: `message_length` bytes.
+///
+/// - SeeAlso:
+/// [gRPC Protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md)
+internal class LengthPrefixedMessageReader {
+  private var buffer: ByteBuffer!
+  private var state: State = .expectingCompressedFlag
+  private let mode: Mode
+
+  internal init(mode: Mode) {
+    self.mode = mode
+  }
+
+  internal enum Mode { case client, server }
+
+  private enum State {
+    case expectingCompressedFlag
+    case expectingMessageLength
+    case receivedMessageLength(Int)
+    case willBuffer(requiredBytes: Int)
+    case isBuffering(requiredBytes: Int)
+  }
+
+  /// Reads bytes from the given buffer until it is exhausted or a message has been read.
+  ///
+  /// Length prefixed messages may be split across multiple input buffers in any of the
+  /// following places:
+  /// 1. after the compression flag,
+  /// 2. after the message length flag,
+  /// 3. at any point within the message.
+  ///
+  /// - Note:
+  /// This method relies on state; if a message is _not_ returned then the next time this
+  /// method is called it expect to read the bytes which follow the most recently read bytes.
+  /// If a message _is_ returned without exhausting the given buffer then reading a
```

Comment:
What do you mean with "then reading a different buffer is not an issue."?

Reply:
I meant that once a message has been returned then any buffer can be provided to the reader, regardless of whether the last buffer had been entirely consumed or not. I think just removing that second sentence makes that clear enough though.

Reply:
👍